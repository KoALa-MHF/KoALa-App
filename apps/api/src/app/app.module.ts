import { Module } from '@nestjs/common';

import { GraphQLModule } from '@nestjs/graphql';
import { ApolloDriver, ApolloDriverConfig } from '@nestjs/apollo';

import { AppController } from './app.controller';
import { AppService } from './app.service';
import { TypeOrmModule } from '@nestjs/typeorm';

import { SessionsModule } from './sessions/sessions.module';
import { Session } from './sessions/entities/session.entity';
import { MediaModule } from './media/media.module';
import { Media } from './media/entities/media.entity';
import { MarkersModule } from './markers/markers.module';
import { Marker } from './markers/entities/marker.entity';
import { UserSessionsModule } from './user-sessions/user-sessions.module';
import { UserSession } from './user-sessions/entities/user-session.entity';
import { AnnotationsModule } from './annotations/annotations.module';
import { Annotation } from './annotations/entities/annotation.entity';
import { formatError } from './core/graphql/grapqhl-error';
import { ConfigModule, databaseConfig, mailConfig } from './config/config.module';
import { ToolbarsModule } from './toolbars/toolbars.module';
import { Toolbar } from './toolbars/entities/toolbar.entity';

import { MailerModule } from '@nestjs-modules/mailer';
import { HandlebarsAdapter } from '@nestjs-modules/mailer/dist/adapters/handlebars.adapter';

@Module({
  imports: [
    ConfigModule,
    TypeOrmModule.forRoot({
      type: 'better-sqlite3',
      database: databaseConfig.name,
      entities: [
        Session,
        Media,
        Marker,
        Annotation,
        UserSession,
        Toolbar,
      ],
      synchronize: true,
    }),
    GraphQLModule.forRoot<ApolloDriverConfig>({
      driver: ApolloDriver,
      debug: true,
      playground: true,
      autoSchemaFile: true,
      formatError: formatError,
      introspection: true,
    }),
    MailerModule.forRoot({
      transport: {
        host: mailConfig.host,
        port: mailConfig.port,
        auth: {
          user: mailConfig.user, // generated ethereal user
          pass: mailConfig.password, // generated ethereal password
        },
        // ignoreTLS: true,
        // secure: false,
      },
      defaults: {
        from: mailConfig.from,
      },
      template: {
        dir: __dirname + '/assets/templates',
        adapter: new HandlebarsAdapter(),
        options: {
          strict: true,
        },
      },
    }),
    SessionsModule,
    ToolbarsModule,
    MediaModule,
    MarkersModule,
    UserSessionsModule,
    AnnotationsModule,
  ],
  controllers: [
    AppController,
  ],
  providers: [
    AppService,
  ],
})
export class AppModule {}
