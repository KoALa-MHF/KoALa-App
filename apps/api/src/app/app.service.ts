import { Injectable } from '@nestjs/common';
import { Message } from '@koala-app/api-interfaces';

@Injectable()
export class AppService {
  getData(): Message {
    return { message: 'Welcome to api!' };
  }
}
