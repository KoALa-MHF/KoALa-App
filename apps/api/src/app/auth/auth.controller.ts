import { Controller, Get, Post, Req, Res, UseGuards } from '@nestjs/common';
import { SamlAuthGuard } from '../core/guards/saml-auth.guard';
import { User } from '../users/entities/user.entity';
import { AuthService } from './auth.service';

@Controller('auth')
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  @Get('sso/saml/login')
  @UseGuards(SamlAuthGuard)
  async samlLogin() {
    //this route is handled by passport-saml
    return;
  }

  @Post('sso/saml/ac')
  @UseGuards(SamlAuthGuard)
  async samlAssertionConsumer(@Req() req, @Res() res) {
    const user: User = req.user;
    if (user) {
      console.log('USER', user);
      const authentication = await this.authService.authenticateSamlUser(user);
      res.redirect('/?jwt=' + authentication.accessToken);
    }
  }
}
