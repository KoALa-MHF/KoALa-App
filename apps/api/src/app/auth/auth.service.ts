import { Injectable, UnauthorizedException } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { EntityNotFoundError } from 'typeorm';
import { SessionsService } from '../sessions/sessions.service';
import { CreateUserSessionInput } from '../user-sessions/dto/create-user-session.input';
import { UserSessionsService } from '../user-sessions/user-sessions.service';
import { User } from '../users/entities/user.entity';
import { Authentication } from './models/autentication.model';

@Injectable()
export class AuthService {
  constructor(
    private jwtService: JwtService,
    private readonly sessionService: SessionsService,
    private readonly userSessionService: UserSessionsService
  ) {}

  async authenticateUserSession(code: string): Promise<Authentication> {
    try {
      const userSession = await this.userSessionService.findOneByCode(code);
      console.log('---------> authenticateUserSession', userSession);
      return this.getAuthentication(userSession.userId);
    } catch (error) {
      if (error instanceof EntityNotFoundError) {
        throw new UnauthorizedException();
      }
      throw error;
    }
  }

  async authenticateSession(code: string): Promise<Authentication> {
    try {
      const session = await this.sessionService.findOneByCode(code);

      const userSession = await this.userSessionService.create({
        sessionId: session.id,
        user: {
          email: 'test' + new Date().getTime() + '@koala-app.de',
        },
      } as CreateUserSessionInput);

      return this.getAuthentication(userSession.userId);
    } catch (error) {
      if (error instanceof EntityNotFoundError) {
        throw new UnauthorizedException();
      }
      throw error;
    }
  }

  private async getAuthentication(userId: number): Promise<Authentication> {
    const accessToken = await this.getAccessToken(userId);
    return {
      accessToken,
      userId,
    } as Authentication;
  }

  private async getAccessToken(userId: number): Promise<string> {
    const payload = {
      sub: userId,
    };
    return await this.jwtService.signAsync(payload);
  }
}
