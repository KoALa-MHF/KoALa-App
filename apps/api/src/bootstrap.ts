import { INestApplication } from '@nestjs/common';
import { GraphQLValidationPipe } from './app/core/pipes/graphql-validation.pipe';
import { graphqlUploadExpress } from 'graphql-upload';
import * as MailDev from 'maildev';
import helmet from 'helmet';

export const GLOBAL_PREFIX = 'api';

interface IMailDev {
  getAllEmail(done: (error: Error, emails: Array<object>) => void): void;
  on(eventName: string, callback: (email: object) => void): void;
}

export function setupApplication(app: INestApplication) {
  app.use(helmet({ contentSecurityPolicy: process.env.NODE_ENV === 'production' ? undefined : false }));
  app.useGlobalPipes(new GraphQLValidationPipe());
  app.use(graphqlUploadExpress({ maxFileSize: 100000000, maxFiles: 10 }));
  app.setGlobalPrefix(GLOBAL_PREFIX);
}

export function setupMailDevelopmentServer(): Promise<IMailDev> {
  return new Promise((resolve, reject) => {
    const maildev = new MailDev({});
    maildev.listen((error) => {
      if (error) {
        reject(error);
      } else {
        resolve(maildev);
      }
    });
  });
}
