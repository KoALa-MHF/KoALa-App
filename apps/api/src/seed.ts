import { NestFactory } from '@nestjs/core';
import { SeedModule } from './app/seed/seed.module';
import { SeedService } from './app/seed/seed.service';

async function bootstrap() {
  if (process.env.NODE_ENV !== 'development') {
    throw new Error('Seeding database is only allowed for development');
  }
  const appContext = await NestFactory.createApplicationContext(SeedModule);
  try {
    const seeder = appContext.get(SeedService);
    await seeder.seed();
  } catch (error) {
    console.log('Seeding database failed', error);
  } finally {
    await appContext.close();
  }
}

bootstrap();
