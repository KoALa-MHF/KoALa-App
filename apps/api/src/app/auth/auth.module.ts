import { Module } from '@nestjs/common';
import { JwtModule } from '@nestjs/jwt';
import { PassportModule } from '@nestjs/passport';
import { SessionsModule } from '../sessions/sessions.module';
import { UserSessionsModule } from '../user-sessions/user-sessions.module';
import { AuthResolver } from './auth.resolver';
import { AuthService } from './auth.service';
import { JwtStrategy } from './strategies/jwt.strategy';

@Module({
  imports: [
    PassportModule,
    JwtModule.register({
      secret: 'jWTSecret',
      signOptions: { expiresIn: '24h' },
    }),
    UserSessionsModule,
    SessionsModule,
  ],
  providers: [
    AuthResolver,
    AuthService,
    JwtStrategy,
  ],
  exports: [],
})
export class AuthModule {}
