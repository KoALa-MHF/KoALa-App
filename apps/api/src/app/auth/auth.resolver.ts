import { Resolver, Mutation, Args, ObjectType, Field } from '@nestjs/graphql';
import { AuthService } from './auth.service';
import { AuthenticateSessionInput } from './dto/authenticate-session.input';
import { AuthenticateUserSessionInput } from './dto/authenticate-user-session.input';
import { Authentication } from './models/autentication.model';

@Resolver(() => Authentication)
export class AuthResolver {
  constructor(private readonly authService: AuthService) {}

  @Mutation(() => Authentication)
  authenticateUserSession(
    @Args('authenticateUserSessionInput')
    authenticateUserSessionInput: AuthenticateUserSessionInput
  ) {
    return this.authService.authenticateUserSession(authenticateUserSessionInput.code);
  }

  @Mutation(() => Authentication)
  authenticateSession(
    @Args('authenticateSessionInput')
    authenticateSessionInput: AuthenticateSessionInput
  ) {
    return this.authService.authenticateSession(authenticateSessionInput.code);
  }
}
