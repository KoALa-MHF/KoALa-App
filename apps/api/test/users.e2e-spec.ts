import request, { LEGACY_WEBSOCKET_PROTOCOL, supertestWs } from 'supertest-graphql';
import gql from 'graphql-tag';

import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';

import { AppModule } from '../src/app/app.module';
import { setupApplication } from '../src/bootstrap';
import { AuthGuardMock } from './mocks/guards/AuthGuard.mock';
import { AuthGuard } from '../src/app/core/guards/auth.guard';
import { SeedModule } from '../src/app/seed/seed.module';
import { SeedService } from '../src/app/seed/seed.service';
import { UsersData } from '../src/app/seed/data/users.data';

const QUERY_ME = gql`
  query Me {
    me {
      id
      email
      displayName
    }
  }
`;

const UPDATE_USER = gql`
  mutation UpdateUser($updateUserInput: UpdateUserInput!) {
    updateUser(updateUserInput: $updateUserInput) {
      id
      email
      displayName
    }
  }
`;

const UPDATE_USER_VARIABLES = {
  updateUserInput: {
    displayName: 'update displayName',
  },
};

describe('Users (e2e)', () => {
  let app: INestApplication;

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [
        SeedModule,
        AppModule,
      ],
    })
      .overrideGuard(AuthGuard)
      .useValue(new AuthGuardMock())
      .compile();

    app = moduleFixture.createNestApplication();
    setupApplication(app);
    await app.init();
    const seeder = app.get(SeedService);
    await seeder.seed();
  });

  afterEach(async () => {
    await app.close();
  });

  describe('Query Me', () => {
    it('Not authenticated user should get "Unauthorized" error', async () => {
      const { errors } = await request(app.getHttpServer()).query(QUERY_ME);

      expect(errors).toHaveLength(1);
      expect(errors[0].message).toBe('Unauthorized');
    });

    it('Authenticated user should get the own user data returned', async () => {
      const { data } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionOwner1.id}`, { type: 'bearer' })
        .query(QUERY_ME)
        .expectNoErrors();

      expect(data).toMatchSnapshot();
    });
  });

  describe('Update User', () => {
    it('Not authenticated user should get "Unauthorized" error', async () => {
      const { errors } = await request(app.getHttpServer()).mutate(UPDATE_USER).variables(UPDATE_USER_VARIABLES);

      expect(errors).toHaveLength(1);
      expect(errors[0].message).toBe('Unauthorized');
    });

    it('Authenticated user can update own user data', async () => {
      const { data } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionOwner1.id}`, { type: 'bearer' })
        .mutate(UPDATE_USER)
        .variables(UPDATE_USER_VARIABLES)
        .expectNoErrors();

      expect(data).toMatchSnapshot();
    });

    it('Authenticated user can only update the right fields', async () => {
      const { data } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionOwner1.id}`, { type: 'bearer' })
        .mutate(UPDATE_USER)
        .variables({
          ...UPDATE_USER_VARIABLES,
          email: 'email-cannot-be-updated@test.com',
          id: 2,
        })
        .expectNoErrors();

      expect(data).toMatchSnapshot();
    });
  });
});
