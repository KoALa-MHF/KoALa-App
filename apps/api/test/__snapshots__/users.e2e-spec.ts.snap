// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users (e2e) Query Me Authenticated SAML users should get the own user data returned with role "User" 1`] = `
{
  "me": {
    "displayName": "Session Owner 1",
    "email": "session-owner-1@koala-app.de",
    "id": "1",
    "role": "USER",
  },
}
`;

exports[`Users (e2e) Query Me Authenticated guest users should get the own user data returned with role "Guest" 1`] = `
{
  "me": {
    "displayName": "Session Participant 1",
    "email": "session-participant-1@koala-app.de",
    "id": "3",
    "role": "GUEST",
  },
}
`;

exports[`Users (e2e) Update User Authenticated user can only update the right fields 1`] = `
{
  "updateUser": {
    "displayName": "update displayName",
    "email": "session-owner-1@koala-app.de",
    "id": "1",
  },
}
`;

exports[`Users (e2e) Update User Authenticated user can update own user data 1`] = `
{
  "updateUser": {
    "displayName": "update displayName",
    "email": "session-owner-1@koala-app.de",
    "id": "1",
  },
}
`;
